libpromises/policyserver.c

```diff
@@ -0,0 +1,300 @@
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA
+
+  To the extent this program is licensed as part of the Enterprise
+  versions of CFEngine, the applicable Commercial Open Source License
+  (COSL) may apply to this file if you as a licensee so wish it. See
+  included file COSL.txt.
+*/
+
+#include <policyserver.h>
+
+#include <eval_context.h>
+#include <addr_lib.h>
+#include <communication.h>
+
+#include <assert.h>
+#include <logging.h>
+#include <string_lib.h>
+#include <file_lib.h>
+
+//*******************************************************************
+// POLICY SERVER VARIABLES:
+//*******************************************************************
+
+static char *POLICY_SERVER      = NULL;
+static char *POLICY_SERVER_HOST = NULL;
+static char  POLICY_SERVER_IP[CF_MAX_IP_LEN];
+static char  POLICY_SERVER_PORT[CF_MAX_PORT_LEN];
+
+
+//*******************************************************************
+// POLICY SERVER SET FUNCTION:
+//*******************************************************************
+
+/* NULL is an acceptable value for new_policy_server. Could happen when an
+ * already bootstrapped server re-parses its policies, and the
+ * policy_server.dat file has been removed. Then this function will be called
+ * with NULL as new_policy_server, and cf-serverd will keep running even
+ * without a policy server set. */
+void SetPolicyServer(EvalContext *ctx, const char *new_policy_server)
+{
+    // Clean up static variables:
+    free(POLICY_SERVER);
+    free(POLICY_SERVER_HOST);
+    POLICY_SERVER         = NULL;
+    POLICY_SERVER_HOST    = NULL;
+
+    POLICY_SERVER_IP[0]   = '\0';
+    POLICY_SERVER_PORT[0] = '\0';
+
+    // Remove variables if undefined policy server:
+    if (new_policy_server == NULL || new_policy_server[0] == '\0')
+    {
+        EvalContextVariableRemoveSpecial(   ctx, SPECIAL_SCOPE_SYS,
+                                            "policy_hub" );
+        EvalContextVariableRemoveSpecial(   ctx, SPECIAL_SCOPE_SYS,
+                                            "policy_hub_ip" );
+        EvalContextVariableRemoveSpecial(   ctx, SPECIAL_SCOPE_SYS,
+                                            "policy_hub_port" );
+        return;
+    }
+
+    // Set POLICY_SERVER full string:
+    POLICY_SERVER = xstrdup(new_policy_server);
+
+    // Parse new_policy_server and set POLICY_SERVER_IP + POLICY_SERVER_PORT
+    {
+        // Parse policy server in a separate buffer:
+        char *host, *port;
+        char buffer[CF_MAX_SERVER_LEN];
+        strncpy(buffer, new_policy_server, CF_MAX_SERVER_LEN);
```

Comment:
`strncpy()` is seldom needed, it's not a good function. Just allocate here, since setting this is not frequent operation.